src/nn/webgl/WebGLModel.js

```diff
@@ -0,0 +1,328 @@
+                 .depthwiseConv2D(filter, [strideH, strideW], 'valid').add(bias)));
+        }
+      } break;
+      case OperationCode.AVERAGE_POOL_2D:
+      case OperationCode.MAX_POOL_2D: {
+        let inCount = inputs.length;
+        if (inCount !== 7 && inCount !== 10) {
+          throw new Error(`Invalid parameters number of Pooling ${op}`);
+        }
+        let i = 0;
+        let input = operands[inputs[i++]];
+        let output = operands[outputs[0]];
+        let strideW, strideH;
+        let filterW, filterH;
+        let activation;
+        if (inCount === 7) {
+          let paddingCode = operands[inputs[i++]].value[0];
+          let padding = PaddingCodeMap.get(paddingCode);
+          strideW = operands[inputs[i++]].value[0];
+          strideH = operands[inputs[i++]].value[0];
+          filterW = operands[inputs[i++]].value[0];
+          filterH = operands[inputs[i++]].value[0];
+          activation = FuseFunctionMap.get(operands[inputs[i++]].value[0]);
+          if (op === OperationCode.AVERAGE_POOL_2D) {
+            output.assign(activation(
+              input.avgPool([filterH, filterW], [strideH, strideW], padding)));
+          } else {
+            output.assign(activation(
+              input.maxPool([filterH, filterW], [strideH, strideW], padding)));
+          }
+        } else {
+          let paddingLeft = operands[inputs[i++]].value[0];
+          let paddingRight = operands[inputs[i++]].value[0];
+          let paddingTop = operands[inputs[i++]].value[0];
+          let paddingBottom = operands[inputs[i++]].value[0];
+          strideW = operands[inputs[i++]].value[0];
+          strideH = operands[inputs[i++]].value[0];
+          filterW = operands[inputs[i++]].value[0];
+          filterH = operands[inputs[i++]].value[0];
+          activation = FuseFunctionMap.get(operands[inputs[i++]].value[0]);
+          if (op === OperationCode.AVERAGE_POOL_2D) {
+            output.assign(activation(
+              input.pad([[0, 0], [paddingTop, paddingBottom], [paddingLeft, paddingRight], [0, 0]])
+                   .avgPool([filterH, filterW], [strideH, strideW], 'valid')));
+          } else {
+            output.assign(activation(
+              input.pad([[0, 0], [paddingTop, paddingBottom], [paddingLeft, paddingRight], [0, 0]])
+                   .maxPool([filterH, filterW], [strideH, strideW], 'valid')));
+          }
+        }
+      } break;
+      case OperationCode.SOFTMAX: {
+        let input = operands[inputs[0]];
+        let beta = operands[inputs[1]].value[0];
+        let output = operands[outputs[0]];
+        output.assign(input.mul(tf.scalar(beta)).softmax());
+      } break;
+      case OperationCode.RESHAPE: {
+        let input = operands[inputs[0]];
+        let targetShape = operands[inputs[1]];
+        let output = operands[outputs[0]];
+        output.assign(input.reshape(targetShape.dataSync()));
```

Comment:
do we need to do `dataSync`? I understand it leads to memory read back from GPU to CPU which is bad for performance.

Reply:
According to the definition of [reshape in NN API](https://github.com/intel/webml-polyfill/blob/master/src/nn/Enums.js#L952), targetShape is a 1-D tensor whose value is stored in GPU. But for [tf.reshape](https://js.tensorflow.org/api/0.13.3/#reshape), target shape should be an array of integers..